### PR TITLE
Update to ASM9 for ABI normalization

### LINKED ISSUE #20394: Support latest JDK features while inspecting classes
Gradle currently uses ASM 9.1 and supports executing on JDK 17.  However, the class inspection used while calculating classes API surface declares that it only supports up to ASM7:  https://github.com/gradle/gradle/blob/3e1d8e2cebe108e657d21c65d6de5cbe511364b6/subprojects/normalization-java/src/main/java/org/gradle/internal/normalization/java/impl/ApiMemberSelector.java#L58.  ASM7 only supports features up to JDK14.

We should upgrade this to ASM9 and address any compatibility issues if they exist.

### Expected Behavior
Classes built with JDK15+ and utilizing features from those newer JDK versions can be inspected

### Current Behavior
The inspector fails with an UnsupportedOperationException when it encounters classes that require a newer version of the ASM api.

### Context
This came up in a build that uses Compile Classpath normalization for an input to a task, which uses class inspection to fingerprint the public API of classes.  The input has a class that uses the JDK15 "sealed" modifier (https://docs.oracle.com/en/java/javase/15/language/sealed-classes-and-interfaces.html).  When the fingerprinting encounters this, it blows up while inspecting the class and throws `java.lang.UnsupportedOperationException: PermittedSubclasses requires ASM9`.  This then causes the normalization to fail, and Gradle ends up falling back to hashing the whole jar file.

### PR DESCRIPTION
**Pull request by octylFractal**
Record components are not carried into ABI as they already create changes via new methods. Permitted subclasses are carred into ABI as they will be able to affect exhaustiveness of switch patterns in the future.

Fixes #20394

### CODE REVIEW

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy [hunk 1]
```diff
@@ -815,6 +815,128 @@ class JavaCompileIntegrationTest extends AbstractIntegrationSpec {
...
+
+        succeeds ":b:compileJava"
+
+        then:
+        executedAndNotSkipped(":b:compileJava")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/20394")
+    @Requires(TestPrecondition.JDK17_OR_LATER)
+    def "class snapshotting changes when a sealed modifier is changed"() {
```
Comment (ghale) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy:881]: Would it make more sense to put these two tests in `AbstractJavaCompileAvoidanceIntegrationSpec`?  (And if so, we should change the test descriptions to match the style in those classes - something like "recompiles when a sealed modifier is changed" or some such)

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy [hunk 2]
```diff
@@ -815,6 +815,128 @@ class JavaCompileIntegrationTest extends AbstractIntegrationSpec {
...
+'''
+
+        succeeds ":b:compileJava"
+
+        then:
+        executedAndNotSkipped(":b:compileJava")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/20394")
+    @Requires(TestPrecondition.JDK17_OR_LATER)
```
Comment (ghale) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy:880]: Aren't sealed modifiers a JDK15 feature?  Or is there some other compatibility issue here I'm not aware of?

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy [hunk 3]
```diff
@@ -815,6 +815,128 @@ class JavaCompileIntegrationTest extends AbstractIntegrationSpec {
         !file("build/classes/java/main/ignored/io/ignored/IgnoredExample.class").exists()
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/20394")
+    @Requires(TestPrecondition.JDK16_OR_LATER)
```
Comment (ghale) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy:819]: Aren't records a JDK14 feature?

**Review by ghale — APPROVED** [review]
A couple of minor questions/suggestions on the tests, but I think this looks good.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy (hunk 3 above)
Comment (octylFractal) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy:819]: They are a JDK 14 _preview_ feature, but only stabilized in 16 [[source]](https://openjdk.java.net/projects/jdk/16/).

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy (hunk 2 above)
Comment (octylFractal) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy:880]: They are a JDK 15 _preview_ feature, but only stabilized in 17 [[source]](https://openjdk.java.net/projects/jdk/17/). I can add special-casing to compile with `--preview` if we want, when needed, but I think the tests are sufficient as they are.

Comment (ghale) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy:880]: Nope, that makes sense.  I wasn't aware these were only preview.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy (hunk 3 above)
Comment (ghale) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy:819]: same here

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy (hunk 1 above)
Comment (octylFractal) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileIntegrationTest.groovy]: That's a much better spot, thanks. I've done this.

**Comment by octylFractal** [comment]
@bot-gradle test and merge

**Pull request merged**